From 2ac1111d62cdfa5523f61ffee9723d722e5c5615 Mon Sep 17 00:00:00 2001
From: "Damien.Horsley" <Damien.Horsley@imgtec.com>
Date: Mon, 13 Jul 2015 16:17:09 +0100
Subject: [PATCH 05/25] clk: pistachio: Add audio pll rates and set parent rate
 for event timer dividers

Change the audio pll to variable and create an audio pll
rate table with two entries

The event timer path contains two dividers. Allow set parent
rate on the second divider to enable these to be calculated
automatically by the clock framework

Change-Id: I35f9b5184e12b6b0dd9d4e8071cc53db7626818f
Signed-off-by: Damien.Horsley <Damien.Horsley@imgtec.com>
---
 drivers/clk/pistachio/clk-pistachio.c | 32 ++++++++++++++++++++++++++++----
 1 file changed, 28 insertions(+), 4 deletions(-)

--- a/drivers/clk/pistachio/clk-pistachio.c
+++ b/drivers/clk/pistachio/clk-pistachio.c
@@ -92,8 +92,8 @@ static struct pistachio_div pistachio_di
 	DIV(CLK_SPI1_DIV, "spi1_div", "spi1_internal_div", 0x254, 7),
 	DIV(CLK_EVENT_TIMER_INTERNAL_DIV, "event_timer_internal_div",
 	    "event_timer_mux", 0x258, 3),
-	DIV(CLK_EVENT_TIMER_DIV, "event_timer_div", "event_timer_internal_div",
-	    0x25c, 12),
+	DIV_F(CLK_EVENT_TIMER_DIV, "event_timer_div", "event_timer_internal_div",
+	    0x25c, 12, CLK_SET_RATE_PARENT, 0),
 	DIV(CLK_AUX_ADC_INTERNAL_DIV, "aux_adc_internal_div",
 	    "aux_adc_internal", 0x260, 3),
 	DIV(CLK_AUX_ADC_DIV, "aux_adc_div", "aux_adc_internal_div", 0x264, 10),
@@ -174,11 +174,35 @@ static struct pistachio_pll_rate_table w
 	},
 };
 
+static struct pistachio_pll_rate_table audio_pll_rates[] = {
+	{
+		.fref		= 52000000,
+		.fout_min	= 146250000,
+		.fout		= 147456000,
+		.fout_max	= 149499999,
+		.refdiv		= 0x1,
+		.fbdiv		= 0x2d,
+		.frac		= 0x5efee6,
+		.postdiv1	= 0x4,
+		.postdiv2	= 0x4,
+	}, {
+		.fref		= 52000000,
+		.fout_min	= 133250000,
+		.fout		= 135475200,
+		.fout_max	= 136500000,
+		.refdiv		= 0x1,
+		.fbdiv		= 0x29,
+		.frac		= 0xaf46fd,
+		.postdiv1	= 0x4,
+		.postdiv2	= 0x4,
+	},
+};
+
 static struct pistachio_pll pistachio_plls[] __initdata = {
 	PLL(CLK_MIPS_PLL, "mips_pll", "xtal", PLL_GF40LP_LAINT, 0x0,
 	    mips_pll_rates),
-	PLL_FIXED(CLK_AUDIO_PLL, "audio_pll", "audio_refclk_mux",
-		  PLL_GF40LP_FRAC, 0xc),
+	PLL(CLK_AUDIO_PLL, "audio_pll", "audio_refclk_mux", PLL_GF40LP_FRAC,
+	    0xc, audio_pll_rates),
 	PLL_FIXED(CLK_RPU_V_PLL, "rpu_v_pll", "xtal", PLL_GF40LP_LAINT, 0x20),
 	PLL_FIXED(CLK_RPU_L_PLL, "rpu_l_pll", "xtal", PLL_GF40LP_LAINT, 0x2c),
 	PLL_FIXED(CLK_SYS_PLL, "sys_pll", "xtal", PLL_GF40LP_FRAC, 0x38),
